File: problem_set/trapping-rain-water/trapping-rain-water.java

```
// @Title: 接雨水 (Trapping Rain Water)
// @Author: robert.sunq
// @Date: 2021-07-01 22:00:29
// @Runtime: 1 ms
// @Memory: 38.2 MB

class Solution {
    public int trap(int[] height) {
        int len = height.length;
        int[] left = new int[len];
        // 记录到当前位置的左侧最大值
        for(int i =0 ;i<len;i++){
            if(i == 0){
                left[i] = height[i];
            }else{
                left[i] = Math.max(left[i-1],height[i]);
            }
        }
        int res = 0;
        int temp = 0;
        int h = 0;
        for(int i = len-1;i>0;i--){
            if(i == len-1){
                temp = height[i];
            }else{
                // 记录到当前位置的右侧最大值
                temp = Math.max(temp,height[i]);
                // 可以接水的最大高度为左右侧的最小值
                h = Math.min(left[i],temp);
                if(h>height[i]){
                    res = res + (h - height[i]);
                }
            }
        }

        return res;
    }
}

File: problem_set/trapping-rain-water/trapping-rain-water.cpp

```
// @Title: 接雨水 (Trapping Rain Water)
// @Author: robert.sunq
// @Date: 2020-05-13 14:54:01
// @Runtime: 8 ms
// @Memory: 6.8 MB

class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        if (n<3) return 0; 
        vector<int> left(n,0),right(n,0);
        for(int i=1;i<n-1;i++){
            left[i] = max(left[i-1],height[i-1]);
        }
        for(int i=n-2;i>=0;i--){
            right[i] = max(right[i+1],height[i+1]);
        }
        int sum = 0;
        for(int i=1;i<n-1;i++){
            int temp = min(right[i],left[i]);
            if(temp > height[i])
                sum = sum+temp-height[i];
        }
        return sum;
    }
};
